blocks/graph.py

```diff
@@ -205,18 +212,11 @@ def replace(self, replacements):
         replacement_keys_cur = []
         replacement_vals_cur = []
         # Sort `replacements` in topological order
-        for node in apply_nodes_sorted:
-            for input_ in node.inputs:
-                if input_ in replacements:
-                    if input_ not in replacement_keys_cur:
-                        replacement_keys_cur.append(input_)
-                        replacement_vals_cur.append(replacements[input_])
-
-        # Add outputs of the computation graph
-        for output in self.outputs:
-            if output in replacements:
-                replacement_keys_cur.append(output)
-                replacement_vals_cur.append(replacements[output])
+        # variables in self.variables are in topological order
+        for variable in self.variables:
+            if variable in replacements:
```

Comment:
You could move you first loop here. Make a copy of replacements and pop items when found. If it's not empty after you print a warning for each missing variable.